flycheck.el

```diff
@@ -5625,129 +5625,172 @@ about TSLint."
+      (push (flycheck-error-new-at
+             ;; We have no specific position to attach the error to, so
+             ;; let's use the top of the file.
+             1 1
+             error-level
+             error-message
+             :id error-code
+             :checker checker
+             :buffer buffer)
+            errors))
     (nreverse errors)))
 
+(defun flycheck-parse-rust-filter-json (output)
+  "Return parsed JSON data from OUTPUT generated by Cargo or rustc.
+
+When JSON output is requested, Cargo and rustc both emit a mix of
+plain text lines and JSON lines.  This function ignores the plain
+text lines, and parses the JSON lines before returning them in a
+list."
+  (let ((json-array-type 'list)
+        (json-false nil)
+        ;; Skip the plain text lines in OUTPUT, keep the JSON lines.
+        (json-lines (seq-filter (lambda (line)
+                                  (string-match-p "^{" line))
+                                (split-string output "\n"))))
+    ;; Each JSON line is a JSON object.
+    (seq-map #'json-read-from-string json-lines)))
+
+(defun flycheck-parse-rustc (output checker buffer)
+  "Parse rustc errors from OUTPUT and return a list of `flycheck-error'.
+
+CHECKER and BUFFER denote the CHECKER that returned OUTPUT and
+the BUFFER that was checked respectively.
+
+The expected format for OUTPUT is a mix of plain text lines and
+JSON lines.  This function ignores the plain text lines and
+parses only JSON lines.  Each JSON line is expected to be a JSON
+object that corresponds to a diagnostic from the compiler.  The
+expected diagnostic format is described there:
+
+https://github.com/rust-lang/rust/blob/master/src/libsyntax/json.rs#L67-L139"
+  (seq-mapcat (lambda (msg)
+                (flycheck-parse-rustc-diagnostic msg checker buffer))
+              (flycheck-parse-rust-filter-json output)))
+
+(defun flycheck-parse-cargo-rustc (output checker buffer)
+  "Parse Cargo errors from OUTPUT and return a list of `flycheck-error'.
+
+CHECKER and BUFFER denote the CHECKER that returned OUTPUT and
+the BUFFER that was checked respectively.
+
+The expected format for OUTPUT is a mix of plain text lines and
+JSON lines.  This function ignores the plain text lines and
+parses only JSON lines.  Each JSON line is expected to be a JSON
+object that represents a message from Cargo.  The format of
+messages emitted by Cargo is described there:
+
+https://github.com/rust-lang/cargo/blob/master/src/cargo/util/machine_message.rs#L20-L31"
+  (let ((errors))
+    (dolist (msg (flycheck-parse-rust-filter-json output))
+      (let-alist msg
+        ;; Errors and warnings from rustc are wrapped by cargo, so we filter and
+        ;; unwrap them, and delegate the actual construction of `flycheck-error'
+        ;; objects to `flycheck-parse-rustc-diagnostic'.
+        (when (string= .reason "compiler-message")
+          (setq errors (nconc errors (flycheck-parse-rustc-diagnostic .message checker buffer))))))
```

Comment:
Good catch.  And here I was using `nconc` because `append` would have been quadratic... but somehow I thought setting the cdr in `nconc` was constant, not linear.  But looking at the source it's definitely not.